[bin/upgrade] display changelog diff before pulling changes

das7pad opened the pull request:
## Description
<!-- Goal of the pull request -->

For https://github.com/overleaf/issues/issues/3737

This PR enhances the `bin/upgrade` script with displaying of changelog entries.

Download (fetch) and applying (pull) of changes is split into two stages now.
We need to fetch the changelog, ask for user confirmation of the changes and only
 then run the actual pull.

See this demo terminal session:

```
toolkit$ git reset --soft origin/master
toolkit$ bin/upgrade 
Warning: current branch is not master, 'jpa-display-changelog' instead
Checking for code update...
No code update available for download
Changelog:
----------
+## TBD
+### Added
+- `bin/upgrade` displays any changes to the changelog and requires extra
+   confirmation before applying the remote changes to the local branch.
+
----------
Perform code update? [y/n] y
Current commit is 0e38cf1
Updating code...
From github.com:overleaf/toolkit
 * branch            jpa-display-changelog -> FETCH_HEAD
Updating 0e38cf1..45368bb
Fast-forward
 CHANGELOG.md |  5 +++++
 bin/upgrade  | 62 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++-----
 2 files changed, 62 insertions(+), 5 deletions(-)
...
```

## Related issues / Pull Requests
<!-- Fixes #xyz, Contributes to #xyz, Related to #xyz-->

For https://github.com/overleaf/issues/issues/3737

## Contributor Agreement

- [x] I confirm I have signed the [Contributor License Agreement](https://github.com/overleaf/overleaf/blob/master/CONTRIBUTING.md#contributor-license-agreement)

ShaneKilkelly commented on bin/upgrade:
```diff
@@ -126,17 +163,32 @@ function handle_git_update() {
+    echo "Code update available for download!"
+
+    local should_fetch="n"
+    read -r -p "Proceed to download update? [y/n] " should_fetch
```
I think these two prompts could be confusing for users, as they might not appreciate the difference between fetch and pull in git-land. Plus, prompting for the initial fetch feels a bit redundant.

I suggest we:
- automatically fetch from origin
- display the changelog
- _prompt_ the user, ask if they want to merge the changes

If (at any stage) there are no changes to pull, we can just stop with an appropriate message.

Otherwise, I think this is a great improvement :tada:

das7pad replied:
> I suggest we:
> 
> * automatically fetch from origin
> * display the changelog
> * _prompt_ the user, ask if they want to merge the changes

Done in https://github.com/overleaf/toolkit/pull/27/commits/4b48169a56e6531d1cd55334150b874f16f7119d

> If (at any stage) there are no changes to pull, we can just stop with an appropriate message.

The user could dismiss the apply step (downloaded changes on remote branch). We should check for local changes either way (as is currently).

das7pad replied:
... and thanks for the feedback :)

ShaneKilkelly reviewed (changes requested):
Looks good overall, one comment about the prompts.

ShaneKilkelly commented:
For anyone curious as to what this looks like in action:
![ezgif com-optimize](https://user-images.githubusercontent.com/1540054/100212838-c1575480-2f05-11eb-8bf4-90c581920a27.gif)

It's quite slick

das7pad commented:
> Ah, one more thing, I notice that `bin/dev/lint` is showing a few warnings (some of which probably pre-date this work). Would you mind fixing them? Thanks!

These are existing errors and all of them relate to declaring and assigning variables in a single statement. I can file a followup PR.

<details>

```
toolkit$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
toolkit$ bin/dev/lint | wc -l
40
toolkit$ git checkout jpa-display-changelog 
Switched to branch 'jpa-display-changelog'
Your branch is up to date with 'origin/jpa-display-changelog'.
toolkit$ bin/dev/lint | wc -l
40
toolkit$ bin/dev/lint | grep -e ' SC'
        ^-- SC2155: Declare and assign separately to avoid masking return values.
        ^-- SC2155: Declare and assign separately to avoid masking return values.
            ^-- SC2155: Declare and assign separately to avoid masking return values.
        ^-- SC2155: Declare and assign separately to avoid masking return values.
        ^-- SC2155: Declare and assign separately to avoid masking return values.
        ^-- SC2155: Declare and assign separately to avoid masking return values.
        ^-- SC2155: Declare and assign separately to avoid masking return values.
        ^-- SC2155: Declare and assign separately to avoid masking return values.
``` 
</details>

ShaneKilkelly reviewed (approved):
Looks good :tada:

ShaneKilkelly commented:
cc @mserranom
